tracee-rules/input.go

```diff
@@ -4,70 +4,174 @@ import (
 	"bufio"
 	"encoding/gob"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"io"
 	"log"
 	"os"
-	"time"
+	"strings"
 
 	"github.com/aquasecurity/tracee/tracee-rules/types"
 	tracee "github.com/aquasecurity/tracee/tracee/external"
 )
 
-func setupStdinSource(inputSource string) (chan types.Event, error) {
-	res := make(chan types.Event)
-	scanner := bufio.NewScanner(os.Stdin)
-	go func() {
-		for scanner.Scan() {
-			event := scanner.Bytes()
-			switch inputSource {
-			case "tracee":
-				var e tracee.Event
-				err := json.Unmarshal(event, &e)
-				if err != nil {
-					log.Printf("invalid json in %s: %v", string(event), err)
-				}
-				res <- types.Event(e)
-			}
-		}
-		close(res)
-	}()
-	return res, nil
+var errHelp = errors.New("user has requested help text")
+
+type inputFormat uint8
+
+const (
+	invalidInputFormat inputFormat = iota
+	jsonInputFormat
+	gobInputFormat
+)
+
+type traceeInputOptions struct {
+	inputFile   *os.File
+	inputFormat inputFormat
 }
 
-func setupTraceeSource(traceeFilePath string) (chan types.Event, error) {
-	_, err := os.Stat(traceeFilePath)
-	if err != nil {
-		return nil, fmt.Errorf("invalid Tracee input file: %s", traceeFilePath)
+func setupTraceeInputSource(opts *traceeInputOptions) (chan types.Event, error) {
+
+	if opts.inputFormat == jsonInputFormat {
+		return setupTraceeJSONInputSource(opts)
 	}
-	f, err := os.Open(traceeFilePath)
-	if err != nil {
-		return nil, fmt.Errorf("invalid file: %s", traceeFilePath)
+
+	if opts.inputFormat == gobInputFormat {
+		return setupTraceeGobInputSource(opts)
```

Comment:
nit: Replace with a switch statement.

Reply:
If it's all the same to you, I think this looks cleaner than this:

```